source/Game.cpp

```diff
@@ -6,35 +6,88 @@
 /* Basic implementation of chess. No players, no checkmate, etc. */
 void Game::run()
 {
-
-	// TODO: Make turn-based
-	// TODO: Check game state (check, stalemate, etc.)
-
+	// TODO: Check checkmate (i.e. no moves for the player in check) and stalemate (no moves for anyone)
+	// Set up turn-based game
 	std::string input;
+	int currentMove = 1;
+	Color toMove = getTurn(currentMove);
 
-	board.printBoardAlgebraicAxes();
-	std::cout << "Enter a space-separated move (like e6 e5): ";
-	std::getline(std::cin, input);
+	// Print instructions
+	std::cout << "Welcome to Chara Chess. Enter moves as space-separated strings (e.g. e2 e4) or exit to quit." << '\n';
 
-	while (input != "-1")
+	// Repeat until game is finished
+	while (true)
 	{
+		// Print board
+		board.printBoardAlgebraicAxes();
+
+		// Get first input
+		std::cout << printColor(toMove) << "'s turn. Enter move: ";
+		std::getline(std::cin, input);
+
+		// Handle non-move commands
+		if (input == "exit")
+		{
+			break;
+		}
+		else if (input == "moves")
+		{
+			board.printMoves();
+			continue;
+		}
+		else if (input == "revert")
+		{
+			if (currentMove > 1)
+			{
+				board.revertLastMove();
+				currentMove--;
+				toMove = getTurn(currentMove);
+			}
+			else
+			{
+				std::cout << "Can't revert - already at beginning of game." << '\n';
+			}
+			continue;
+		}
+		else if (input == "captures")
+		{
+			board.printCapturedPieces();
+			continue;
+		}
+
+		// Convert input to a move
 		std::string buf;					// for parsing input
 		std::stringstream ss(input);       	// insert the string into a stream
 		std::vector<std::string> tokens; 	// create vector to hold substrings
-
 		while (ss >> buf)
 		{
 			tokens.push_back(buf);
 		}
-
 		std::pair<int, int> from = board.algebraicToInt(tokens[0]); 	// convert substring to pair
 		std::pair<int, int> to = board.algebraicToInt(tokens[1]);		// convert substring to pair
 
+		// Verify that the correct player is moving
+		if (board.getPiece(from)->getColor() != toMove)
```

Comment:
```suggestion
		if (board.getPiece(from) != NULL)
		{
			if (board.getPiece(from)->getColor() != toMove)
			{
				std::cout << "Error: It's " << printColor(toMove) << "'s turn." << '\n';
				continue;
			}
		}
		else
		{
			std::cout << "Error: No piece to move." << "\n";
		}
```

Reply:
Something like this, to allow me to input an invalid move without throwing a nullptr exception.

Reply:
Let me know if my most recent commit doesn't resolve this for you. I never had any exceptions, but I also never put in malformed commands. Obviously this has no error checking whatsoever, so anything that's outsides bounds of accepted input will crash program.